src/FileFormats/NL/NL.jl

```diff
@@ -0,0 +1,788 @@
+end
+
+const _SCALAR_FUNCTIONS = Union{
+    MOI.SingleVariable,
+    MOI.ScalarAffineFunction{Float64},
+    MOI.ScalarQuadraticFunction{Float64},
+}
+
+const _SCALAR_SETS = Union{
+    MOI.LessThan{Float64},
+    MOI.GreaterThan{Float64},
+    MOI.EqualTo{Float64},
+    MOI.Interval{Float64},
+}
+
+function MOI.supports_constraint(
+    ::Model,
+    ::Type{<:_SCALAR_FUNCTIONS},
+    ::Type{<:_SCALAR_SETS},
+)
+    return true
+end
+
+function MOI.supports_constraint(
+    ::Model,
+    ::Type{MOI.SingleVariable},
+    ::Type{<:Union{MOI.ZeroOne,MOI.Integer}},
+)
+    return true
+end
+
+MOI.supports(::Model, ::MOI.ObjectiveSense) = true
+MOI.supports(::Model, ::MOI.ObjectiveFunction{<:_SCALAR_FUNCTIONS}) = true
+
+# ==============================================================================
+
+function MOI.supports(
+    ::Model,
+    ::MOI.VariablePrimalStart,
+    ::Type{MOI.VariableIndex},
+)
+    return true
+end
+
+function MOI.set(model::Model, ::MOI.VariablePrimalStart, x, v::Real)
+    model.x[x].start = Float64(v)
+    return
+end
+
+function MOI.set(model::Model, ::MOI.VariablePrimalStart, x, ::Nothing)
+    model.x[x].start = nothing
+    return
+end
+
+function MOI.get(model::Model, ::MOI.VariablePrimalStart, x::MOI.VariableIndex)
+    return model.x[x].start
+end
+
+# ==============================================================================
+
+struct _LinearNLPEvaluator <: MOI.AbstractNLPEvaluator end
+MOI.features_available(::_LinearNLPEvaluator) = [:ExprGraph]
+MOI.initialize(::_LinearNLPEvaluator, ::Vector{Symbol}) = nothing
+
+function MOI.copy_to(
+    dest::Model,
+    model::MOI.ModelLike;
+    copy_names::Bool = false,
+)
+    mapping = MOI.Utilities.IndexMap()
+    # Initialize the NLP block.
+    nlp_block = try
+        MOI.get(model, MOI.NLPBlock())
+    catch
+        MOI.NLPBlockData(MOI.NLPBoundsPair[], _LinearNLPEvaluator(), false)
+    end
+    if !(:ExprGraph in MOI.features_available(nlp_block.evaluator))
+        error(
+            "Unable to use AmplNLWriter because the nonlinear evaluator " *
+            "does not supply expression graphs.",
+        )
+    end
+    MOI.initialize(nlp_block.evaluator, [:ExprGraph])
+    # Objective function.
+    if nlp_block.has_objective
+        dest.f = _NLExpr(MOI.objective_expr(nlp_block.evaluator))
+    else
+        F = MOI.get(model, MOI.ObjectiveFunctionType())
+        obj = MOI.get(model, MOI.ObjectiveFunction{F}())
+        dest.f = _NLExpr(obj)
+    end
+    # Nonlinear constraints
+    for (i, bound) in enumerate(nlp_block.constraint_bounds)
+        push!(
+            dest.g,
+            _NLConstraint(MOI.constraint_expr(nlp_block.evaluator, i), bound),
+        )
+    end
+    dest.nlpblock_dim = length(dest.g)
+    starts = MOI.supports(model, MOI.VariablePrimalStart(), MOI.VariableIndex)
+    for x in MOI.get(model, MOI.ListOfVariableIndices())
+        dest.x[x] = _VariableInfo()
```

Comment:
Since `x` is build like this, if `model` has variables from 1 to n, `x` will have it too. It therefore make sense to use `CleverDict`

Reply:
I think it's fine as-is. There isn't much to be gained from a clever dict, and we don't need the ordered nature of the original variables.